kotlin/workflow-testing/src/main/java/com/squareup/workflow/testing/WorkflowTester.kt

```diff
@@ -231,3 +183,110 @@ class WorkflowTester<InputT : Any, OutputT : Any, RenderingT : Any> @TestOnly in
+ *
+ * All workflow-related coroutines are cancelled when the block exits.
+ */
+@TestOnly
+fun <T, OutputT : Any, RenderingT : Any> Workflow<Unit, OutputT, RenderingT>.testFromStart(
+  snapshot: Snapshot? = null,
+  context: CoroutineContext = EmptyCoroutineContext,
+  block: WorkflowTester<Unit, OutputT, RenderingT>.() -> T
+): T = testFromStart(Unit, snapshot, context, block)
+
+/**
+ * Creates a [WorkflowTester] to run this workflow for unit testing.
+ * If the workflow is [stateful][StatefulWorkflow], [initialState][StatefulWorkflow.initialState]
+ * is not called. Instead, the workflow is started from the given [initialState].
+ *
+ * All workflow-related coroutines are cancelled when the block exits.
+ */
+// @formatter:off
+@TestOnly
+fun <T, InputT : Any, StateT : Any, OutputT : Any, RenderingT : Any>
+    StatefulWorkflow<InputT, StateT, OutputT, RenderingT>.testFromState(
+      input: InputT,
+      initialState: StateT,
+      context: CoroutineContext = EmptyCoroutineContext,
+      block: WorkflowTester<InputT, OutputT, RenderingT>.() -> T
+    ): T = test(block, context) { factory, inputs ->
+      inputs.offer(input)
+      factory.runTestFromState(this, inputs, initialState)
+    }
+// @formatter:on
+
+/**
+ * Creates a [WorkflowTester] to run this workflow for unit testing.
+ * If the workflow is [stateful][StatefulWorkflow], [initialState][StatefulWorkflow.initialState]
+ * is not called. Instead, the workflow is started from the given [initialState].
+ *
+ * All workflow-related coroutines are cancelled when the block exits.
+ */
+// @formatter:off
+@TestOnly
+fun <StateT : Any, OutputT : Any, RenderingT : Any>
+    StatefulWorkflow<Unit, StateT, OutputT, RenderingT>.testFromState(
+      initialState: StateT,
+      context: CoroutineContext = EmptyCoroutineContext,
+      block: WorkflowTester<Unit, OutputT, RenderingT>.() -> Unit
+    ) = testFromState(Unit, initialState, context, block)
+// @formatter:on
+
+@UseExperimental(InternalCoroutinesApi::class)
+private fun <T, I : Any, O : Any, R : Any> test(
+  testBlock: (WorkflowTester<I, O, R>) -> T,
+  baseContext: CoroutineContext,
+  starter: (WorkflowHost.Factory, inputs: Channel<I>) -> WorkflowHost<I, O, R>
+): T {
+  val context = Dispatchers.Unconfined + baseContext + Job(parent = baseContext[Job])
+  val inputs = Channel<I>(capacity = 1)
+  @Suppress("ReplaceSingleLineLet")
+  val host = WorkflowHost.Factory(context)
+      .let { starter(it, inputs) }
+      .let { WorkflowTester(inputs, it, context) }
+
+  var error: Throwable? = null
+  try {
+    return testBlock(host)
+  } catch (e: Throwable) {
+    error = e
+    throw e
+  } finally {
+    if (error != null) {
+      // TODO https://github.com/square/workflow/issues/188 Stop using parameterized cancel.
+      @Suppress("DEPRECATION")
+      val cancelled = context.cancel(error)
+      if (!cancelled) {
+        val cancellationCause = context[Job]!!.getCancellationException()
```

Comment:
`requireNotNull()` would give better error in case of NPE.

Reply:
It's not possible to have a null here, since we explicitly give the context a `Job` on line 262. If it does have a null, someone edited this file and broke something, and I'm not worried about the error message in that case.